### PR TITLE
refactor(protocol-designer): add madge & remove most circular dependencies

### PR DESCRIPTION
**Pull request by IanLondon**
## overview

We've been having some complex webpack / jest problems with unresolved circular imports, which this PR is intended to address.

1. Adds `madge` so we can check for circular dependencies. To check for circular dependencies in PD, do `yarn madge --circular protocol-designer/src/index.js`.
2. Removes circular dependencies by moving some functions around, detailed in changelog below. If you run madge to check, you should see 1 in this PR, whereas on `edge` there are 10. (Remember to do `make install-js` to update your node_modules)

## next steps

Once I resolve the final circular dependency in PD (`persist.js > tutorial/reducers.js`), I'm planning on adding the madge circular dependency check to our builds (at least for PD, but I imagine we could gradually do the same for the other JS projects) so that these circularities won't keep quietly building up over months until we have to deal with them like this.

## Note: hierarchy

I am not sure anymore what the divide btw these 3 "internal modules" of PD is, and we should clean it up / define it better. But as it stands, the order matters to avoid circularity. It's `steplist <- step-forms <- ui/steps`. Never import in the other direction. So `ui/steps` can import from both of the other 2 modules, and in contrast `steplist` cannot import either of the other 2 modules.

## changelog

* add madge as devDependency; add .madgerc file for manual madge use

### steplist, step-froms, ui/steps
* move 'saveStepForm' thunk from 'steplist' to 'step-forms' to remove circular import
* clean up `ui/steps` exports (which required moving action types into `ui/steps/actions/types.js`)
* move `getCurrentFormCanBeSaved` from `ui/steps` to `step-forms` to remove some circular dependencies
* quick circular import removal: steplist importing thru its own index.js

### step-generation
* remove circular dependencies in `step-generation` by refactoring `utils/index.js` functions into `utils/misc.js` for internal `step-generation` use

## review requests

- [ ] smoke test PD functionality (make some steps, move them around, check browser console for unusual errors)
- [ ] code review

### CODE REVIEW

File: .madgerc [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "detectiveOptions": {
+    "es6": {
+      "skipTypeImports": true
```
Comment (IanLondon) [inline .madgerc]: this prevents madge from counting `import type` imports for Flow as circular imports

File: protocol-designer/src/ui/steps/index.js [hunk 2]
```diff
@@ -1,7 +1,10 @@
 // @flow
 
-import * as actions from './actions'
-
+import * as _actions from './actions/actions'
```
Comment (IanLondon) [inline protocol-designer/src/ui/steps/index.js]: is this a sane way to do the exports? I want the thunks in their own file distinct from plain actions, but when importing any thunks from outside the `ui/steps` inner module,  I want the consumer to do `import {actions as stepsActions} from '../ui/steps'` and not have to import actions vs thunks separately.

**Comment by mcous** [comment]
Added a temporary do-not-merge to be lifted once our upcoming release branch is cut:

- This PR adds JS dependencies to the lockfile
- `electron-builder` has known bugs with dependency resolution that has, in the past, resulted in an unexpected dependency being pulled in and breaking a release

File: protocol-designer/src/ui/steps/index.js (hunk 2 above)
Comment (Kadee80) [reply protocol-designer/src/ui/steps/index.js]: I'm not sure I have a strong opinion yet. Maybe we meet in person tomorrow to discuss the pros and cons. If importing/exporting them separately is a thing that's consistently causing issues, it's worth exploring.

**Pull request merged**